test/integration/eds_integration_test.cc

```diff
@@ -0,0 +1,120 @@
+      socket_address->set_address(Network::Test::getLoopbackAddressString(GetParam()));
+      socket_address->set_port_value(fake_upstreams_[i]->localAddress()->ip()->port());
+      if (i >= healthy_endpoints) {
+        endpoint->set_health_status(envoy::api::v2::core::HealthStatus::UNHEALTHY);
+      }
+    }
+
+    // Write to file the DiscoveryResponse and trigger inotify watch.
+    envoy::api::v2::DiscoveryResponse eds_response;
+    eds_response.set_version_info(std::to_string(eds_version_++));
+    eds_response.set_type_url(Config::TypeUrl::get().ClusterLoadAssignment);
+    eds_response.add_resources()->PackFrom(cluster_load_assignment);
+    // Past the initial write, need move semantics to trigger inotify move event that the
+    // FilesystemSubscriptionImpl is subscribed to.
+    if (eds_path_.empty()) {
+      eds_path_ =
+          TestEnvironment::writeStringToFileForTest("eds.pb_text", eds_response.DebugString());
+    } else {
+      std::string path = TestEnvironment::writeStringToFileForTest("eds.update.pb_text",
+                                                                   eds_response.DebugString());
+      ASSERT_EQ(0, ::rename(path.c_str(), eds_path_.c_str()));
+    }
+  }
+
+  void createUpstreams() override {}
+
+  void initialize() override {
+    updateEndpoints(0, 0);
+    config_helper_.addConfigModifier([this](envoy::config::bootstrap::v2::Bootstrap& bootstrap) {
+      // Switch predefined cluster_0 to EDS filesystem sourcing.
+      auto* cluster_0 = bootstrap.mutable_static_resources()->mutable_clusters(0);
+      cluster_0->mutable_hosts()->Clear();
+      cluster_0->set_type(envoy::api::v2::Cluster::EDS);
+      auto* eds_cluster_config = cluster_0->mutable_eds_cluster_config();
+      eds_cluster_config->mutable_eds_config()->set_path(eds_path_);
+    });
+    HttpIntegrationTest::initialize();
+    for (uint32_t i = 0; i < upstream_endpoints_; ++i) {
+      fake_upstreams_.emplace_back(new FakeUpstream(0, FakeHttpConnection::Type::HTTP1, version_));
+    }
+  }
+
+  static constexpr uint32_t upstream_endpoints_ = 5;
+  std::string eds_path_;
+  uint32_t eds_version_{};
+};
+
+INSTANTIATE_TEST_CASE_P(IpVersions, EdsIntegrationTest,
+                        testing::ValuesIn(TestEnvironment::getIpVersionsForTest()));
+
+// Validate that health status updates are consumed from EDS.
+TEST_P(EdsIntegrationTest, HealthUpdate) {
+  initialize();
+  // Initial state, no cluster members.
+  EXPECT_EQ(1, test_server_->counter("cluster.cluster_0.update_success")->value());
+  EXPECT_EQ(0, test_server_->counter("cluster.cluster_0.membership_change")->value());
+  EXPECT_EQ(0, test_server_->gauge("cluster.cluster_0.membership_total")->value());
+  EXPECT_EQ(0, test_server_->gauge("cluster.cluster_0.membership_healthy")->value());
+  // 2/2 healthy endpoints.
+  updateEndpoints(2, 2);
+  test_server_->waitForCounterGe("cluster.cluster_0.update_success", 2);
```

Comment:
We had a similar pattern with the load tests, and it became problematic as we added more tests as each update required manually increment each integer.  Can you change this either to something like 

waitForCounterGe("...update_success", update_successes++) 

or wrapper the whole thing

s/updateEndpoints/setEndpoints/ and add
void updateEndpoints(int endpoints, int healthy) {
  setEndpoints(endpoints, healthy);
 test_server_->waitForCounterGe(..... update_counter_++);
}